packages/hir/src/solver/types.candy

```diff
@@ -0,0 +1,220 @@
+      }
+      return Some<Map<SolverVariable, SolverType>>(Map.of1<SolverVariable, SolverType>(Tuple(this_, other)))
+    }
+
+    if this is SolverValue && other is SolverVariable {
+      return other.unify(this)
+    }
+
+    let this_ = (this as SolverValue)
+    let other_ = (other as SolverValue)
+
+    if !(this_.name == other_.name)
+      || !((this_.parameters as Iterable).length() == (other_.parameters as Iterable).length()) {
+      return None<Map<SolverVariable, SolverType>>()
+    }
+
+    // To unify all types, we have a queue of substitutions that we try to bring into a common
+    // solution space one by one.
+    let solutionSpace = MutableMap.empty<SolverVariable, SolverType>()
+    let queue: Maybe<MutableList<(SolverVariable, SolverType)>> = (this_.parameters as Iterable<SolverType>).zip<SolverType>(other_.parameters)
+      .map<Maybe<Map<SolverVariable, SolverType>>>({ it.first.unify(it.second) })
+      .foldLeft<Maybe<Iterable<Map<SolverVariable, SolverType>>>>(
+        Some<Iterable<Map<SolverVariable, SolverType>>>(List.empty<Map<SolverVariable, SolverType>>()),
+        { a, b =>
+          if a is None { return None<Iterable<Map<SolverVariable, SolverType>>>() }
+          if b is None { return None<Iterable<Map<SolverVariable, SolverType>>>() }
+          return Some<Iterable<Map<SolverVariable, SolverType>>>(
+            a.unwrap().followedBy(List.of1<Map<SolverVariable, SolverType>>(b.unwrap())),
+          )
+        },
+      )
+      .map<MutableList<(SolverVariable, SolverType)>>({
+        it.expand<(SolverVariable, SolverType)>({ it.entries() }).toMutableList()
+      })
+    if queue is None {
+      return None<Map<SolverVariable, SolverType>>()
+    }
+    let queue = queue.unwrap()
+
+    while (queue as Iterable).isNotEmpty() {
+      let substitution = queue.removeFirst()
+      if (solutionSpace as Map<SolverVariable, SolverType>).containsKey(substitution.first) {
+        let existingSubstitution = (solutionSpace as Map<SolverVariable, SolverType>).get(substitution.first).unwrap()
+        let unified = existingSubstitution.unify(substitution.second)
+        if unified is None {
+          return None<Map<SolverVariable, SolverType>>()
+        }
+        queue.appendAll(unified.unwrap().entries())
+      } else {
+        solutionSpace.set(substitution.first, substitution.second)
+        solutionSpace.updateAll({ it.substitute(substitution.first, substitution.second) })
+        queue.updateAll({
+          Tuple(it.first, it.second.substitute(substitution.first, substitution.second))
+        })
+        if (queue as Iterable<(SolverVariable, SolverType)>).any({ it.second.containsVariable(it.first) }) {
+          return None<Map<SolverVariable, SolverType>>()
+        }
+      }
+    }
+
+    Some<Map<SolverVariable, SolverType>>(solutionSpace)
+  }
+}
+impl SolverType: Hash {
+  fun hash<T>(hasher: Hasher<T>) { assert(false, "subclasses should override this") }
```

Comment:
Can't we just remove this method then and declare `trait SolverType: Hash & Equals`